File: 2. Liquidos/1. Cliente CF 3.5/Oxigenos.Nucleo/CargaDatosForm.vb
Imports OpenNETCF.Net
Imports OpenNETCF.WindowsCE
Imports System.Net
Imports MovilidadCF.Windows.Forms

Public Class CargaDatosForm
    Implements IEstadoCarga

    Private m_ws As OxigenosWS.DataAccess

    Private m_Thread As OpenNETCF.Threading.Thread2
    Private m_IdProceso As Integer = ProcesosComunicacion.CargaCompleta
    Private m_bUpdateCurrentRows As Boolean = False
    Private m_DescripProceso As String = "Descarga datos"
    Private m_sData As String = Nothing
    Private m_bCancelado As Boolean = False
    Private m_bUsarGPRS As Boolean = False
    Private Delegate Sub ActualizarMensajeHandler(ByVal sData As String)
    Private Delegate Sub SetProgresoHandler(ByVal Progreso As Integer)


    Public Shared Function Run(ByVal IdProceso As Integer, ByVal sNombreProceso As String, _
    ByVal bUpdateCurrentRows As Boolean, ByVal bUsarGPRS As Boolean) As Boolean

        Dim result As Boolean = False

        ' Si asi se especifica se intenta primero establecer la conexión GPRS
        UIHandler.StartWait()
        If bUsarGPRS Then
            If Not GPRS.Conectar() Then
                UIHandler.EndWait()
                Return False
            End If
        End If

        ' Se crea la instancia de la forma y se inicia el proceso
        Dim frm As New CargaDatosForm
        frm.m_IdProceso = IdProceso
        frm.m_DescripProceso = sNombreProceso
        frm.m_bUpdateCurrentRows = bUpdateCurrentRows
        frm.m_bUsarGPRS = bUsarGPRS
        result = (UIHandler.ShowDialog(frm) = System.Windows.Forms.DialogResult.OK)
        frm.Dispose()
        ' Finalmente se cierra la conexión
        If bUsarGPRS Then
            GPRS.Desconectar()
        End If
        UIHandler.EndWait()
        Return result
    End Function


    Private Sub CargaDatosForm_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
        ' Se inicializan controles
        UiHandler1.Parent = Me
        pnlInfoTablas.Visible = False
        Me.lblProceso.Text = Me.m_DescripProceso

        ' Se inicia el hilo de ejecución de la integración de datos y se habilita el timer
        ' de reset de estado Idle del sistema
        timerResetIdleState.Enabled = True
        m_Thread = New OpenNETCF.Threading.Thread2(AddressOf IntegrarDatosThread)
        m_Thread.Start()
        UIHandler.EndWait()
    End Sub

    Private Sub ConfigurarWebService()
        m_ws = New OxigenosWS.DataAccess

        Dim prefijo As String = "http://"

        If Settings.UsaHTTPS Then
            prefijo = "https://"
        End If

        If m_bUsarGPRS Then
            m_ws.Url = prefijo & Settings.IPServidorGPRS & ":" & Settings.PuertoServidor & "/" & _
                Settings.VirtualDirectory & "/DataAccess.asmx"
        Else
            m_ws.Url = prefijo & Settings.IPServidor & ":" & Settings.PuertoServidor & "/" & _
                Settings.VirtualDirectory & "/DataAccess.asmx"
        End If
        m_ws.Timeout = TIMEOUT_WEBSERVICE
        If Settings.UsuarioWebService <> "" Then
            m_ws.Credentials = New NetworkCredential(Settings.UsuarioWebService, Settings.ClaveWebService)
        End If
    End Sub

    Private Sub IntegrarDatosThread()
        Dim sFecha As String
        Try
            IniciarObtenerDatos("Obteniendo datos del servidor...")
            ConfigurarWebService()

            ' Se sincroniza la hora del sistema con la hora del servidor
            m_ws.Timeout = 10000
            sFecha = m_ws.GetFechaSistema()
            DateTimeHelper.LocalTime = DateTime.ParseExact(sFecha, "dd/MM/yyyy HH:mm:ss", Nothing)
            m_ws.Timeout = TIMEOUT_WEBSERVICE

            m_sData = m_ws.GetDatosCarga(Settings.IDTerminal, Me.m_IdProceso, Programa.Codigo)
            IniciarIntegracionDatos("Integrando datos...")
            If Not Me.Cancelado Then
                GestorNucleo.IntegrarDatos(Me.m_sData, Me.m_bUpdateCurrentRows, Me)
            End If
            m_ws.ConfirmarCarga(Settings.IDTerminal, Me.m_IdProceso, Programa.Codigo)
            MostrarProcesoTerminado("Proceso de carga de datos terminado exitosamente")
        Catch ex As Exception
            MostrarError(ex.GetType().ToString & vbCrLf & ex.Message)
            WriteLog(ex)
        End Try
    End Sub

    Private Sub InternalMostrarProcesoTerminado(ByVal sMensaje As String)
        pnlInfoTablas.Visible = False
        pnlTerminado.Visible = True
        pnlTerminado.BringToFront()
        lblMensajeFinal.Text = sMensaje
        btnTerminar.Focus()
    End Sub

    Private Sub MostrarProcesoTerminado(ByVal sMensaje As String)
        Dim funcHandler As New ActualizarMensajeHandler(AddressOf InternalMostrarProcesoTerminado)
        Me.Invoke(funcHandler, sMensaje)
    End Sub

    Private Sub IniciarObtenerDatos(ByVal sMensaje As String)
        Dim funcHandler As New ActualizarMensajeHandler(AddressOf InternalIniciarObtenerDatos)
        Me.Invoke(funcHandler, sMensaje)
    End Sub

    Private Sub IniciarIntegracionDatos(ByVal sMensaje As String)
        Dim funcHandler As New ActualizarMensajeHandler(AddressOf InternalIniciarIntegracionDatos)
        Me.Invoke(funcHandler, sMensaje)
    End Sub

    Private Sub InternalIniciarObtenerDatos(ByVal sMensaje As String)
        lblFase.Text = sMensaje
        pnlInfoTablas.Visible = False
        pnlObtenerDatos.Visible = True
        AnimateCtl1.StartAnimation()
    End Sub

    Private Sub InternalIniciarIntegracionDatos(ByVal sMensaje As String)
        lblFase.Text = sMensaje
        AnimateCtl1.StopAnimation()
        pnlObtenerDatos.Visible = False
        pnlInfoTablas.Visible = True
    End Sub

    Private Sub MostrarError(ByVal sMensaje As String)
        Dim funcHandler As New ActualizarMensajeHandler(AddressOf InternalMostrarError)
        Me.Invoke(funcHandler, sMensaje)
    End Sub

    Private Sub InternalMostrarError(ByVal sMensaje As String)
        pnlError.Visible = True
        pnlError.BringToFront()
        lblError.Text = sMensaje
        btnCerrarByError.Focus()
    End Sub

    Private Sub timerResetIdleState_Tick(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles timerResetIdleState.Tick
        OpenNETCF.WindowsCE.PowerManagement.ResetSystemIdleTimer()
    End Sub

    Private Sub btnTerminar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnTerminar.Click
        UIHandler.StartWait()
        timerResetIdleState.Enabled = False
        DialogResult = System.Windows.Forms.DialogResult.OK
    End Sub

    Private Sub btnCancelar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCancelar.Click
        m_bCancelado = True
        lblFase.Text = "Cancelando..."
        pnlInfoTablas.Visible = False
        pbAvanceTotal.Visible = False
        btnCancelar.Enabled = False
        Application.DoEvents()
        If m_Thread IsNot Nothing AndAlso m_Thread.State <> OpenNETCF.Threading.ThreadState.Stopped Then
            If Not m_Thread.Join(5000) Then
                Try
                    m_Thread.Abort()
                Catch
                End Try
            End If
        End If
        UIHandler.StartWait()
        timerResetIdleState.Enabled = False
        DialogResult = System.Windows.Forms.DialogResult.Cancel
    End Sub

    Private Sub btnCerrarByError_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCerrarByError.Click
        UIHandler.StartWait()
        timerResetIdleState.Enabled = False
        DialogResult = System.Windows.Forms.DialogResult.Cancel
    End Sub

    Public Shared Sub WriteLog(ByVal ex As Exception)
        ' Se guarda la información del error en el archivo
        Dim stream As New System.IO.StreamWriter("\ErroresIntegracion.TXT", True)
        Dim bld As New System.Text.StringBuilder()
        Dim inner As Exception = ex.InnerException

        stream.WriteLine("==============================================================================")
        stream.WriteLine(Now.ToString() & "Ha ocurrido una excepción: " & ex.GetType().FullName)
        stream.WriteLine(ex.Message)
        If Not inner Is Nothing Then
            stream.WriteLine("Inner Exception: " & inner.ToString())
        End If
        If ex.GetType() Is GetType(System.Data.SqlServerCe.SqlCeException) Then
            Dim sqlex As System.Data.SqlServerCe.SqlCeException
            Dim err As System.Data.SqlServerCe.SqlCeError
            sqlex = DirectCast(ex, System.Data.SqlServerCe.SqlCeException)

            ' Enumerate each error to a message box.
            For Each err In sqlex.Errors
                stream.WriteLine(" Error Code: " & err.HResult.ToString("X"))
                stream.WriteLine(" Message   : " & err.Message)
                stream.WriteLine(" Minor Err.: " & err.NativeError)
                stream.WriteLine(" Source    : " & err.Source)

                ' Retrieve the error parameter numbers for each error.
                Dim numPar As Integer
                For Each numPar In err.NumericErrorParameters
                    If 0 <> numPar Then
                        stream.WriteLine(" Num. Par. : " & numPar)
                    End If
                Next numPar

                ' Retrieve the error parameters for each error.
                Dim errPar As String
                For Each errPar In err.ErrorParameters
                    If String.Empty <> errPar Then
                        bld.Append((ControlChars.Cr & " Err. Par. : " & errPar))
                    End If
                Next errPar
            Next err
        End If
        stream.WriteLine("Stack Trace: " & ex.StackTrace.ToString())
        stream.Close()
    End Sub


#Region " Implementación IEstadoCarga "

    Public ReadOnly Property Cancelado() As Boolean Implements IEstadoCarga.Cancelado
        Get
            Return m_bCancelado
        End Get
    End Property

    Private Sub InternalIniciarTabla(ByVal sNombreTabla As String)
        pbAvanceTabla.Value = 0
        lblTabla.Text = "Procesando " & sNombreTabla
    End Sub

    Private Sub SetProgresosTabla(ByVal Progreso As Integer)
        pbAvanceTabla.Value = Progreso
    End Sub

    Private Sub SetProgresoTotal(ByVal Progreso As Integer)
        pbAvanceTotal.Value = Progreso
    End Sub

    Public Sub IniciarTabla(ByVal sNombreTabla As String) Implements IEstadoCarga.IniciarTabla
        Dim funcHandler As New ActualizarMensajeHandler(AddressOf Me.InternalIniciarTabla)
        Me.Invoke(funcHandler, sNombreTabla)
    End Sub

    Public WriteOnly Property ProgresoTabla() As Integer Implements IEstadoCarga.ProgresoTabla
        Set(ByVal value As Integer)
            Dim funcHandler As New SetProgresoHandler(AddressOf Me.SetProgresosTabla)
            Me.Invoke(funcHandler, value)
        End Set
    End Property

    Public WriteOnly Property ProgresoTotal() As Integer Implements IEstadoCarga.ProgresoTotal
        Set(ByVal value As Integer)
            Dim funcHandler As New SetProgresoHandler(AddressOf Me.SetProgresoTotal)
            Me.Invoke(funcHandler, value)
        End Set
    End Property

#End Region

End Class

File: 1. Herramientas/4. Praxair Comunicaciones Gases/Nueva/Comunicacion/CargaDatosForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using MovilidadCF.Windows.Forms;
using System.Net;
using OpenNETCF.Net;
using OpenNETCF.WindowsCE;
using MovilidadCF.Data;
using MovilidadCF.Compression;


namespace PraxairComunicaciones.Comunicacion
{
    internal partial class CargaDatosForm : Form, MovilidadCF.Data.DataSetSerializer.IEstadoCarga
    {
        //private QuickMobile.Client.WSComunicaciones.DataAccess m_ws;
        private WSPraxair.DataAccess m_ws;
        private OpenNETCF.Threading.Thread2 m_Thread;
        private string m_sData = null;
        private bool m_bCancelado = false;
        private delegate void ActualizarMensajeHandler(string sData);
        private delegate void SetProgresoHandler(int Progreso);
        private bool m_bCerrarVentana = false;        
        private ParametrosWS m_rwParametros;
        private static ConfiguracionServidorWS m_ConfiguracionWS;
        private string m_sConexionLocal = null;

        public CargaDatosForm()
        {
            InitializeComponent();
        }

        public static void ConfiguracionServidorWS(string pIPServidor, string pPuertoServidor, string pVirtualDirectory, string pUsuarioWebService, string pClaveWebService, int pTimeOutWebServices, string pIPServidorGPRS, bool usaHTTPS)
        {
            m_ConfiguracionWS = new ConfiguracionServidorWS(pIPServidor, pPuertoServidor, pVirtualDirectory, pUsuarioWebService, pClaveWebService, pTimeOutWebServices,pIPServidorGPRS, usaHTTPS);
        }        

        public static bool Run(int sPrograma, int sProceso, string sIdTerminal, string sIdSistema, bool bComprimir, bool bUpdateCurrentRows, bool bUsarGPRS, bool bCerrarVentana, string sNombreProceso, string sConexionGPRS, string sConexionLocal,MovilidadCF.Data.DataSetSerializer.QueryList ListQuery, params object[] rParametros)
        {

            bool result = false;
            bUsarGPRS = usarGprsf(sConexionGPRS);
            // Si asi se especifica se intenta primero establecer la conexión GPRS
            UIHandler.StartWait();
            
            if (bUsarGPRS)
            {
                if (!(GPRS.Conectar(sConexionGPRS)))
                {
                    UIHandler.EndWait();
                    return false;
                }
            }
            
            // Controlar las Excepciones
            try
            {
                
                // Se crea la instancia de la forma y se inicia el proceso
                CargaDatosForm frm = new CargaDatosForm();
                frm.m_rwParametros = new ParametrosWS(sPrograma, sProceso, sIdTerminal, sIdSistema, bComprimir, bUpdateCurrentRows, bUsarGPRS, sNombreProceso, sConexionGPRS, ListQuery, rParametros);
                frm.m_bCerrarVentana = bCerrarVentana;
                frm.m_sConexionLocal = sConexionLocal;
                result = (UIHandler.ShowDialog(frm) == System.Windows.Forms.DialogResult.OK);
                //result = (frm.ShowDialog() == System.Windows.Forms.DialogResult.OK);
                frm.Dispose();
                // Finalmente se cierra la conexión
                if (bUsarGPRS)
                {
                    GPRS.Desconectar();
                }
                return result;
            }
            catch
            {
                return false;
            }
            finally
            {
                UIHandler.EndWait();
            }
        }

        #region Miembros de IEstadoCarga

        public bool Cancelado
        {
            get { return m_bCancelado; }
        }

        public int ProgresoTabla
        {
            set
            {
                SetProgresoHandler funcHandler = new SetProgresoHandler(this.SetProgresosTabla);
                this.Invoke(funcHandler, value);
            }
        }

        public int ProgresoTotal
        {
            set
            {
                SetProgresoHandler funcHandler = new SetProgresoHandler(this.SetProgresoTotal);
                this.Invoke(funcHandler, value);
            }
        }

        public void IniciarTabla(string sNombreTabla)
        {
            ActualizarMensajeHandler funcHandler = new ActualizarMensajeHandler(this.InternalIniciarTabla);
            this.Invoke(funcHandler, sNombreTabla);
        }

        private void InternalIniciarTabla(string sNombreTabla)
        {
            pbAvanceTabla.Value = 0;
            lblTabla.Text = "Procesando " + sNombreTabla;
        }

        private void SetProgresosTabla(int Progreso)
        {
            pbAvanceTabla.Value = Progreso;
        }

        private void SetProgresoTotal(int Progreso)
        {
            pbAvanceTotal.Value = Progreso;
        }

        #endregion

        public static  bool usarGprsf( string sConexionGPRS)
        {
            if (sConexionGPRS == null)
                return false;

            if (sConexionGPRS.ToString().Trim().Equals(""))
                return false;

            string prefijo = "http://";

            WSPraxair.DataAccess m_ws;
            m_ws = new PraxairComunicaciones.WSPraxair.DataAccess();

            if (m_ConfiguracionWS.UtilizaHTTPS)
            {
                prefijo = "https://";
                m_ws.Url = prefijo + m_ConfiguracionWS.IPServidor + "/" + m_ConfiguracionWS.VirtualDirectory + "/DataAccess.asmx";
            }
            else
            {
                m_ws.Url = prefijo + m_ConfiguracionWS.IPServidor + ":" + m_ConfiguracionWS.PuertoServidor + "/" + m_ConfiguracionWS.VirtualDirectory + "/DataAccess.asmx";
            }


            MovilidadCF.Logging.Logger.Write(m_ws.Url);

            m_ws.Timeout = m_ConfiguracionWS.TimeOutWebServices;
            m_ConfiguracionWS.TimeOutWebServices = m_ws.Timeout;
            m_ws.Timeout = 10000;
            
            try
            {
                string fecha = m_ws.GetFechaSistema();
                return false;
            }
            catch(Exception ex)
            {
                
                return true;
            }
        }

        private void CargaDatosForm_Load(object sender, EventArgs e)
        {
            // Se inicializan controles
            UiHandler1.Parent = this;
            pnlInfoTablas.Visible = false;
            this.lblProceso.Text = this.m_rwParametros.DescripcionProceso;

            // Se inicia el hilo de ejecución de la integración de datos y se habilita el timer
            // de reset de estado Idle del sistema
            timerResetIdleState.Enabled = true;
            m_Thread = new OpenNETCF.Threading.Thread2(IntegrarDatosThread);
            m_Thread.Start();
            this.CargarImagenes();
            UIHandler.EndWait();
        }

        private void ConfigurarWebService()
        {
            string prefijo = "http://";

            if (m_ConfiguracionWS.UtilizaHTTPS)
            {
                prefijo = "https://";
            }

            //m_ws = new QuickMobile.Client.WSComunicaciones.DataAccess();
            m_ws = new PraxairComunicaciones.WSPraxair.DataAccess();
            if (this.m_rwParametros.UsarGPRS)
            {
                
                m_ws.ConnectionGroupName = this.m_rwParametros.ConexionGPRS  ;

                if (m_ConfiguracionWS.UtilizaHTTPS)
                {
                    prefijo = "https://";
                    m_ws.Url = prefijo + m_ConfiguracionWS.IPServidor + "/" + m_ConfiguracionWS.VirtualDirectory + "/DataAccess.asmx";
                }
                else
                {
                    m_ws.Url = prefijo + m_ConfiguracionWS.IPServidor + ":" + m_ConfiguracionWS.PuertoServidor + "/" + m_ConfiguracionWS.VirtualDirectory + "/DataAccess.asmx";
                }              
            }
            else
            {
                if (m_ConfiguracionWS.UtilizaHTTPS)
                {
                    prefijo = "https://";
                    m_ws.Url = prefijo + m_ConfiguracionWS.IPServidor + "/" + m_ConfiguracionWS.VirtualDirectory + "/DataAccess.asmx";
                    System.Net.ServicePointManager.CertificatePolicy = (ICertificatePolicy)new TrustAllCertificatePolicy();
                }
                else
                {
                    m_ws.Url = prefijo + m_ConfiguracionWS.IPServidor + ":" + m_ConfiguracionWS.PuertoServidor + "/" + m_ConfiguracionWS.VirtualDirectory + "/DataAccess.asmx";
                }
            }

            MovilidadCF.Logging.Logger.Write(m_ws.Url);
            m_ws.Timeout = m_ConfiguracionWS.TimeOutWebServices;
            m_ConfiguracionWS.TimeOutWebServices = m_ws.Timeout;

            
            if (m_ConfiguracionWS.UsuarioWebService != "")
            {
                m_ws.Credentials = new NetworkCredential(m_ConfiguracionWS.UsuarioWebService, m_ConfiguracionWS.ClaveWebService);
            }
        }

        private void IntegrarDatosThread()
        {
            string sFecha = null;
            try
            {
                IniciarObtenerDatos("Obteniendo datos del servidor...");
                ConfigurarWebService();                
                // Se sincroniza la hora del sistema con la hora del servidor
                m_ws.Timeout = 10000;
                sFecha = m_ws.GetFechaSistema();
                DateTimeHelper.LocalTime = DateTime.ParseExact(sFecha, "dd/MM/yyyy HH:mm:ss", null);
                m_ws.Timeout = m_ConfiguracionWS.TimeOutWebServices;
                m_ConfiguracionWS.TimeOutWebServices = m_ws.Timeout;
                string sIdComunicaciones = "";
                //m_sData = m_ws.CargarDatos(m_rwParametros.Programa, m_rwParametros.Proceso, m_rwParametros.IdTerminal, m_rwParametros.IdSistema, m_rwParametros.Comprimir, ref sIdComunicaciones, m_rwParametros.Parametros);
                m_sData = m_ws.GetDatosCarga(m_rwParametros.IdTerminal, m_rwParametros.Proceso, m_rwParametros.Programa);
                m_rwParametros.IdComunicaciones = sIdComunicaciones;
                
                IniciarIntegracionDatos("Integrando datos...");
                if (!this.Cancelado)
                {
                    //m_GestorIntegracion.LoadPrimaryKeysInfo();
                    DataSetSerializer dsDatos = new DataSetSerializer(new MovilidadCF.Data.SqlServerCe.GestorDatosBase(m_sConexionLocal));                    
                     
                    if (m_rwParametros.Comprimir)
                    {
                        string datos;
                        //datos = MovilidadCF.Compression.DataCompression.UnCompress(Convert.FromBase64String(m_sData), Convert.FromBase64String(m_sData).Length);
                        //DateTime.ParseExact("2010/06/16 00:00:00", "yyyy/MM/dd hh:mm:ss", null);
                       
                        dsDatos.SaveCompressToDatabase(this.m_sData, m_rwParametros.UpdateCurrentRows, this, null);
                    }
                    else
                    {
                        dsDatos.SaveToDatabase(this.m_sData, m_rwParametros.UpdateCurrentRows, this, null);                    
                    }                                        
                }
                m_ws.ConfirmarCarga(m_rwParametros.IdTerminal, m_rwParametros.Proceso, m_rwParametros.Programa);
                MostrarProcesoTerminado("Proceso de carga de datos terminado exitosamente");
            }
            catch (Exception ex)
            {
                if (m_rwParametros.IdComunicaciones != null)
                {
                    //CONTROLAR POR QUE EL CATCH PUEDE SER POR COMUNICACION Y SE BLOQUEA
                    try
                    {
                        m_ws.ConfirmarCarga(m_rwParametros.IdTerminal, m_rwParametros.Proceso, m_rwParametros.Programa);
                    }
                    catch (Exception ex2)
                    {
                        WriteLog(ex2);
                    }
                }
                MostrarError(ex.GetType().ToString() + "\r\n" + ex.Message);
                WriteLog(ex);
            }
        }

        private void InternalMostrarProcesoTerminado(string sMensaje)
        {
            pnlInfoTablas.Visible = false;
            pnlTerminado.Visible = true;
            pnlTerminado.BringToFront();
            lblMensajeFinal.Text = sMensaje;
            btnTerminar.Focus();
            if (m_bCerrarVentana)
            {
                this.btnTerminar_Click(this, null);
            }
        }

        private void MostrarProcesoTerminado(string sMensaje)
        {
            ActualizarMensajeHandler funcHandler = new ActualizarMensajeHandler(InternalMostrarProcesoTerminado);
            this.Invoke(funcHandler, sMensaje);
        }

        private void IniciarObtenerDatos(string sMensaje)
        {
            ActualizarMensajeHandler funcHandler = new ActualizarMensajeHandler(InternalIniciarObtenerDatos);
            this.Invoke(funcHandler, sMensaje);
        }

        private void IniciarIntegracionDatos(string sMensaje)
        {
            ActualizarMensajeHandler funcHandler = new ActualizarMensajeHandler(InternalIniciarIntegracionDatos);
            this.Invoke(funcHandler, sMensaje);
        }

        private void InternalIniciarObtenerDatos(string sMensaje)
        {
            lblFase.Text = sMensaje;
            pnlInfoTablas.Visible = false;
            pnlObtenerDatos.Visible = true;
            AnimateCtl1.StartAnimation();
        }

        private void InternalIniciarIntegracionDatos(string sMensaje)
        {
            lblFase.Text = sMensaje;
            AnimateCtl1.StopAnimation();
            pnlObtenerDatos.Visible = false;
            pnlInfoTablas.Visible = true;
        }

        private void MostrarError(string sMensaje)
        {
            ActualizarMensajeHandler funcHandler = new ActualizarMensajeHandler(InternalMostrarError);
            this.Invoke(funcHandler, sMensaje);
        }

        private void InternalMostrarError(string sMensaje)
        {
            pnlError.Visible = true;
            pnlError.BringToFront();
            lblError.Text = sMensaje;
            btnCerrarByError.Focus();
        }

        private void timerResetIdleState_Tick(object sender, EventArgs e)
        {
            OpenNETCF.WindowsCE.PowerManagement.ResetSystemIdleTimer();
        }

        private void btnTerminar_Click(object sender, EventArgs e)
        {
            UIHandler.StartWait();
            timerResetIdleState.Enabled = false;
            DialogResult = System.Windows.Forms.DialogResult.OK;
        }

        private void btnCancelar_Click(object sender, EventArgs e)
        {
            m_bCancelado = true;
            lblFase.Text = "Cancelando...";
            pnlInfoTablas.Visible = false;
            pbAvanceTotal.Visible = false;
            btnCancelar.Enabled = false;
            Application.DoEvents();
            if (m_Thread != null && m_Thread.State != OpenNETCF.Threading.ThreadState.Stopped)
            {
                if (!(m_Thread.Join(5000)))
                {
                    try
                    {
                        m_Thread.Abort();
                    }
                    catch
                    {
                    }
                }
            }
            UIHandler.StartWait();
            timerResetIdleState.Enabled = false;
            DialogResult = System.Windows.Forms.DialogResult.Cancel;
        }

        private void btnCerrarByError_Click(object sender, EventArgs e)
        {
            UIHandler.StartWait();
            timerResetIdleState.Enabled = false;
            DialogResult = System.Windows.Forms.DialogResult.Cancel;
        }

        public static void WriteLog(Exception ex)
        {
            // Se guarda la información del error en el archivo
            System.IO.StreamWriter stream = new System.IO.StreamWriter("\\ErroresIntegracion.TXT", true);
            System.Text.StringBuilder bld = new System.Text.StringBuilder();
            Exception inner = ex.InnerException;

            stream.WriteLine("==============================================================================");
            stream.WriteLine(System.DateTime.Now.ToString() + "Ha ocurrido una excepción: " + ex.GetType().FullName);
            stream.WriteLine(ex.Message);
            if (inner != null)
            {
                stream.WriteLine("Inner Exception: " + inner.ToString());
            }
            if (ex.GetType() == typeof(System.Data.SqlServerCe.SqlCeException))
            {
                System.Data.SqlServerCe.SqlCeException sqlex = null;
                //INSTANT C# NOTE: Commented this declaration since looping variables in 'foreach' loops are declared in the 'foreach' header in C#:
                //			System.Data.SqlServerCe.SqlCeError err = null;
                sqlex = (System.Data.SqlServerCe.SqlCeException)ex;

                // Enumerate each error to a message box.
                foreach (System.Data.SqlServerCe.SqlCeError err in sqlex.Errors)
                {
                    stream.WriteLine(" Error Code: " + err.HResult.ToString("X"));
                    stream.WriteLine(" Message   : " + err.Message);
                    stream.WriteLine(" Minor Err.: " + err.NativeError);
                    stream.WriteLine(" Source    : " + err.Source);

                    // Retrieve the error parameter numbers for each error.
                    //INSTANT C# NOTE: Commented this declaration since looping variables in 'foreach' loops are declared in the 'foreach' header in C#:
                    //				int numPar = 0;
                    foreach (int numPar in err.NumericErrorParameters)
                    {
                        if (0 != numPar)
                        {
                            stream.WriteLine(" Num. Par. : " + numPar);
                        }
                    }

                    // Retrieve the error parameters for each error.
                    //INSTANT C# NOTE: Commented this declaration since looping variables in 'foreach' loops are declared in the 'foreach' header in C#:
                    //				string errPar = null;
                    foreach (string errPar in err.ErrorParameters)
                    {
                        if (string.Empty != errPar)
                        {
                            bld.Append(("\r" + " Err. Par. : " + errPar));
                        }
                    }
                }
            }
            stream.WriteLine("Stack Trace: " + ex.StackTrace.ToString());
            stream.Close();
        }

        private void CargarImagenes()
        {
            System.Reflection.Assembly _assembly = System.Reflection.Assembly.GetExecutingAssembly();
            this.AnimateCtl1.Image = new System.Drawing.Bitmap(_assembly.GetManifestResourceStream("PraxairComunicaciones.Imagenes.AnimateCtl1.Image.jpg"));              
        }

    }
}